treewide: Disable automatic argparse argument shortening

nordicjm opened the pull request:
Disables allowing the python argparse library from automatically shortening command line arguments, this prevents issues whereby a new command is added and code that wrongly uses the shortened command of an existing argument which is the same as the new command being added will silently change script behaviour.

Fixes #53495

- [X] Scripts updated
- [X] Script invocations updated
- [x] Release notes added
- [X] RFC treewide change email sent
- [X] Architecture working group discussed
- [X] pylint check added to ensure additional python code isn't added that uses argparse without this
- [x] Have pylint check run in CI

nordicjm committed:
treewide: Disable automatic argparse argument shortening

Disables allowing the python argparse library from automatically
shortening command line arguments, this prevents issues whereby
a new command is added and code that wrongly uses the shortened
command of an existing argument which is the same as the new
command being added will silently change script behaviour.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm commented:
@stephanosio would it be possible to run a full CI on this (the weekly one) just to check as best as possible that it hasn't left anything broken?

stephanosio commented:
> @stephanosio would it be possible to run a full CI on this (the weekly one) just to check as best as possible that it hasn't left anything broken?

One of the reasons that we run "full CI" on bi-weekly basis rather than daily or per PR is that it is very expensive to run.

Unless there is a very good reason as to why this PR needs to be treated specially, the full CI run should be run on schedule after it is merged -- if anything is broken, either this can be reverted or hotfixes can be provided.

nordicjm commented:
> Unless there is a very good reason as to why this PR needs to be treated specially, the full CI run should be run on schedule after it is merged -- if anything is broken, either this can be reverted or hotfixes can be provided.

I think code in zephyr is OK, but what I'm concerned about is scripts in and/or being called by other repos e.g. HAL modules, can wait for a full run though.

nordicjm commented:
Compliance issues are from existing code which is unchanged in this PR.

carlescufi commented:
Architecture WG:

- Requires an addition in the release notes
- Requires a pylint custom rule to verify this during compliance on PR
- No objections during the meeting

nordicjm committed:
scripts: pylint: Add argument parser abbreviation checker

Adds a pylint checker for ensuring that the argument parser
library is setup with abbreviations disabled.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm commented:
Added a pylint checker for this, but unsure of how to call it, according to the docs https://pylint.pycqa.org/en/latest/development_guide/how_tos/custom_checkers.html it needs to either be a full python package or in the PYTHON_PATH - so how would the python path be updated for the CI build to include the additional directory?
Also used 99 for the base group because their documentation just says it needs to be unique and not conflict with any of their base groups without listing what they are, so 99 seems like a safe bet.

nordicjm committed:
scripts: ci: Enable pylint check for argument parser abbreviations

Enables running CI checks which will now search for initialisation
of argument parser where abbreviations are not disabled.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
doc: release-notes 3.3: Add note on python argument parser change

Adds a note that python command line argument abbreviations have
been disabled, are no longer allowed in future and instructions of
how to update out-of-tree scripts that used abbreviated arguments.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

de-nordic reviewed (dismissed):
migrate_mcmgr_kconfigs.py looks OK.

fabiobaltieri commented on scripts/ci/check_compliance.py:
```diff
@@ -826,14 +830,23 @@ def run(self):
 
-        pylintcmd = ["pylint", "--rcfile=" + pylintrc] + py_files
+        python_environment = os.environ.copy()
+        if "PYTHONPATH" in python_environment:
+            python_environment["PYTHONPATH"] = check_script_dir + ":" + \
+                                               python_environment["PYTHONPATH"]
+        else:
+            python_environment["PYTHONPATH"] = check_script_dir
```
You could always set and only handle the extend case:

```
python_environment["PYTHONPATH"] = check_script_dir 
if "PYTHONPATH" in os.environ:
  python_environment["PYTHONPATH"] += ":" + os.environ['PYTHONPATH"]
```

saves an "else", up to you really.

nordicjm commented:
@fabiobaltieri can you merge this overriding the compliance check?

fabiobaltieri commented:
> @fabiobaltieri can you merge this overriding the compliance check?

Nope, looks like it's marked as required, I have no permissions to bypass it. @carlescufi can you do it?
